Adding memory profiling page. (#236)
micro-ROS/micro-ROS.github.io: _docs/concepts/memo_prof/index.md

Finally notice that the nature and size of the memory consumed by publications and subscriptions is quite different. The reason is that subscribers stockpile the data in pre-allocated memory buffers. This enables storage of messages in slots (whose number depends on the XRCE-DDS and RMW histories) for a twofold purpose: to avoid loss of data in the case Reliable communication has been opted for, and this is reflected by the XRCE-DDS history, and to allow a flexible message handling and asynchronous message passing between the various layers, which is reflected in the RMW_history. The case of services matches closely that of subscribers, since in a request/response pattern all entities involved need to receive, and thus store and handle, data.

The diagrams shown below address all the features discussed in this section in a graphical way.

> ![Pubsub diagram](pubsub_diagram.png)

Reviewer: Should this image to be centered like the next one in line 67?

> ![Pubsub diagram](pubsub_diagram.png)

Reviewer: ```suggestion
![Pub/sub diagram](pubsub_diagram.png)
```

<p align="center">
  Fig. 1: Illustrative diagram of the memory management of the micro-ROS Client library in publishers and subscribers applications.
</p>

> <img alt="Services diamgram" src="services_diagram.png" class="center" width="50%">

Reviewer: ```suggestion
<img alt="Services diagram" src="services_diagram.png" class="center" width="50%">
```

<p align="center">
  Fig. 2: Illustrative diagram of the memory management of the micro-ROS Client library in service applications.
</p>

### Memory buffers

> The total memory consumed by a micro-ROS application running on a MCU (in this specific case, the ESP32 used for testing) can be calculated as the direct sum of different chunks of memory, each devoted to a different function: static memory, stack and dynamic memory.

Reviewer: ```suggestion
The total memory consumed by a micro-ROS application running on a MCU can be calculated as the direct sum of different chunks of memory, each devoted to a different function: static memory, stack and dynamic memory.
```
Use of the ESP32 is explained at the very beginning as the results are probably quite independent of the MCU, I would not repeat it here.

In this section we give a brief description on these kinds of memory used by the micro-ROS library.

*Static Memory*

The static memory has been calculated as the difference between the memory occupied by the .bss and .data sections with a non-zero number of entities, and the memory occupied by the same sections when no micro-ROS application is running, that is, the memory occupied by the rest of components of the RTOS and libraries. This allows discriminating between the memory effectively occupied by micro-ROS and the one that is not specific to it. Notice that we don’t take into consideration neither the constant data stored in flash memory (.text section) nor the data buffer, which is the buffer that stores the data before serialization, since it is specific to the user application and not directly related to the micro-ROS Client operations.

*Stack Memory*

> The stack is the memory consumed by the functions used by the program, when executing. Whilst the static memory can be straightforwardly calculated by just analyzing the compiled binary objects, the stack is the chunk of memory one cannot know precisely before running the application. The stack consumed during the program execution is taken into account by means of a FreeRTOS specific function involved in the memory management capabilities offered by this RTOS, the uxTaskGetStackHighWaterMark() funcion. This function returns the amount of stack that remains unused when the stack consumed by the program is at its greatest value. By subtracting this figure to the total stack available, which is known, one can obtain the stack peak used by the app.

Reviewer: ```suggestion
The stack is the memory consumed by the functions used by the program, when executing. Whilst the static memory can be straightforwardly calculated by just analyzing the compiled binary objects, the stack is the chunk of memory one cannot know precisely before running the application. The stack consumed during the program execution is taken into account by means of a FreeRTOS specific function involved in the memory management capabilities offered by this RTOS, the `uxTaskGetStackHighWaterMark()` function. This function returns the amount of stack that remains unused when the stack consumed by the program is at its greatest value. By subtracting this figure to the total stack available, which is known, one can obtain the stack peak used by the app.
```

*Dynamic Memory*

This is the memory dynamically allocated by the program by calls to `calloc()` and `malloc()` functions in the C language. To measure it we have hijacked the call to dynamic memory related functions since the ROS 2 stack allows users to feed the program with custom memory allocators. 

To better understand micro-ROS’ use of dynamic memory, we need to differentiate between two stages of the micro-ROS operation. In the first stage micro-ROS is initialized, entities such as nodes, publishers and subscribers are created, and all layers get ready for operation. This is the configuration stage and micro-ROS performs all the dynamic memory operations here. The second is the operation stage, in which the actual publications, subscriptions and all other node operations occur. This stage is dynamic memory free in the whole micro-ROS stack. An optional third stage exists, in which the micro-ROS layers are closed and cleaned and all the dynamic memory allocated in the configuration stage is freed.

## Measurements and methodology

### Tested scenarios

> In this section, we summarize the experimental set-up and the different scenarios explored in order to provide a comprehensive review on the memory footprint of the micro-ROS library.

Reviewer: ```suggestion
In this section, we summarize the experimental setup and the different scenarios explored in order to provide a comprehensive review on the memory footprint of the micro-ROS library.
```

In general, our aim is to assess how both the total memory and its independent constituents (static, stack and dynamic) are affected by:

> * The topic size (in the form of an array of bytes of variable size)

Reviewer: Does topic size mean the message size (i.e. the size of the messages on the topic) or is there more to it? If no, I propose to simply write _message size_.

Author: Ok, changed most of them. Sometimes I still refer to "topic" anyways.

* The number of ROS entities (pub/sub and service/client)
* The communication stream type used (Reliable vs Best-Effort)
* The transport protocol (UDP and serial)

> In the first setup, we analyse the total memory consumption of applications that publish or subscribe to topics of variable size while sweeping through the number of entities (publishers and subscribers) and employ UDP transport. We do so for the two different types of communication modes allowed by the middleware library, Reliable and Best-Effort.

Reviewer: ```suggestion
In the first setup, we analyse the total memory consumption of applications that publish or subscribe to topics of variable size while sweeping through the number of entities (publishers and subscribers) and employ UDP transport. We do so for the two different QoS types, Reliable and Best-Effort.
```

In the second setup, we report on how the total memory is distributed between static, stack and dynamic.

In the third setup, we change the transport from UDP to serial, and repeat the calculation for an individual subscription only and for a fixed topic size.

The fourth set of measurements was taken for one subscription only, for a fixed topic size and varying the history cache of the RMW layer from 1 to 20 units.

In the fifth set of measurements we measure the footprint of applications of requesters/repliers that act according to a client/service pattern.

### Results

In this section, we detail the methodology employed for the memory profiling of the experimental configurations described above, and for each of them we present the results obtained.

The measurements are conducted on a micro-ROS Client application with a varying number of entities: either publishers/subscribers (1, 5, 10, 15) or client/server (1, 2, 5, 10).

> All the tested apps run on top of FreeRTOS and inside of an ESP32 board. The board is connected by either UDP or serial transport to a micro-ROS Agent running on a Linux machine. As explained above, the choice of FreeRTOS has been by virtue of its memory management functionalities, which easily allow to compute the memory used by applications.

Reviewer: Probably the ESP32 can then be omitted in the abstract.

Author: I think it's normal to repeat something that is already said in the Abstract, so I'd leave that. But I definitely took away its previous occurrence.

In order to provide an assessment as much realistic as possible, the following parameters have been set to their default values: the creation mode employed was by XML in all tested cases, the MTU was held fixed to its default value of 512 B, and the XRCE-DDS library history cache was always kept fixed to 4.

#### Pub-Sub apps

*Total memory as a function of entities number and topic size*

In this section, we report the total memory used by either publisher or subscriber applications in both Best-Effort and Reliable modes, using UDP transport, an RMW history of 4 slots, an MTU of 512 B and an XRCE-DDS history of 4 slots. The total memory consumption is reported as a function of the entity number and topic size.

> The number of publishers/subscribers has been varied between 1, 5, 10 and 15. Notice that this is equivalent to changing the number of topics, since in our design of the set-up we associate each publisher/subscriber with just one topic.

Reviewer: The numbers 1, 5, 10, and 15 were already given a few paragraphs before. This repetition should be removed in my opinion.

> In principle, in the Reliable case one can occupy the generated buffers with topic sizes up to MTU * XRCE_history, whereas in the Best-Effort case it can be filled with topics up to MTU, which correspond respectively to to 512 B * 4 = 2048 B and 512 B with our default chosen values. This is due to the absence of fragmentation in Best-Effort communication streams, while, thanks to fragmentation, an entity functioning in Reliable mode can send/receive a message opportunely chunked in a number of pieces equal to the XRCE_history, each of the size of the MTU. However, from table 1 one can see the topic size only ranges from 0 and 1366 B in the case of Reliable entities, and between 0 and 490 B for entities in Best-Effort mode. This is due to the fact that in both cases some memory is consumed by headers and, most importantly, in the Reliable case some is consumed by confirmation messages such as heartbeats and acknacks.

Reviewer: ```suggestion
In principle, in the Reliable case one can occupy the generated buffers with topic sizes up to MTU &#215; XRCE_history, whereas in the Best-Effort case it can be filled with topics up to MTU, which correspond respectively to to 512 B &#215; 4 = 2048 B and 512 B with our default chosen values. This is due to the absence of fragmentation in Best-Effort communication streams, while, thanks to fragmentation, an entity functioning in Reliable mode can send/receive a message opportunely chunked in a number of pieces equal to the XRCE_history, each of the size of the MTU. However, from table 1 one can see the topic size only ranges from 0 and 1366 B in the case of Reliable entities, and between 0 and 490 B for entities in Best-Effort mode. This is due to the fact that in both cases some memory is consumed by headers and, most importantly, in the Reliable case some is consumed by confirmation messages such as heartbeats and acknacks.
```
